File: in_Java/0234 Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        int l = 0;
        ListNode d = head;
        while(d != null){
            l++;
            d = d.next;
        }
        if(l <= 1) return true;
        int dis = 0;
        if(l % 2 == 0) dis = l/2;
        else dis = l/2+1;
        ListNode newH = head;
        while(dis != 0){
            newH = newH.next;
            dis--;
        }
        ListNode b = rev(newH);
        ListNode a = head;
        while(b != null){
            if(a.val != b.val) return false;
            a = a.next;
            b = b.next;
        }
        return true;
    }
    private ListNode rev(ListNode n){
        ListNode pre = null;
        ListNode cur = n;
        ListNode next = null;
        while(cur != null){
            next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
}

File: in_Python/0234 Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rev(self, n):
        pre = None
        cur = n
        nextn = None
        while cur != None:
            nextn = cur.next
            cur.next = pre
            pre = cur
            cur = nextn
        return pre
            
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        L = 0
        h = head
        while h != None:
            h = h.next
            L += 1
        mid = L/2
        newH = head
        while mid != 0:
            newH = newH.next
            mid -= 1
        if L % 2 == 1:
            newH = newH.next
        mid = L/2
        
        nh = self.rev(newH)
        while mid != 0:
            if head.val != nh.val:
                return False
            head = head.next
            nh = nh.next
            mid -= 1
        return True
    